homeassistant/components/smappee/__init__.py

```diff
@@ -65,17 +73,32 @@ async def async_setup(hass: HomeAssistant, config: dict):
 
 
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
-    """Set up Smappee from a config entry."""
-    implementation = await config_entry_oauth2_flow.async_get_config_entry_implementation(
-        hass, entry
-    )
+    """Set up Smappee from a zeroconf or config entry."""
+    if CONF_IP_ADDRESS in entry.data:
+        smappee_api = api.api.SmappeeLocalApi(ip=entry.data[CONF_IP_ADDRESS])
+        smappee = Smappee(api=smappee_api, serialnumber=entry.data[CONF_SERIALNUMBER])
+        await hass.async_add_executor_job(smappee.load_local_service_location)
+    else:
+        implementation = await config_entry_oauth2_flow.async_get_config_entry_implementation(
+            hass, entry
+        )
+
+        smappee_api = api.ConfigEntrySmappeeApi(hass, entry, implementation)
 
-    smappee_api = api.ConfigEntrySmappeeApi(hass, entry, implementation)
+        smappee = Smappee(api=smappee_api)
+        await hass.async_add_executor_job(smappee.load_service_locations)
 
-    smappee = Smappee(smappee_api)
-    await hass.async_add_executor_job(smappee.load_service_locations)
+        if not entry.unique_id:
+            serialnumbers = "_".join(
+                [s.device_serial_number for s in smappee.service_locations.values()]
+            )
+            hass.config_entries.async_update_entry(
+                entry,
+                unique_id=f"SmappeeCloud{entry.entry_id}",
```

Comment:
But the serialnumbers are only available if the integration has been setup completely? So should we change the cloud unique_id to a general one, instead of a id that contain the serialnumber(s)?

Reply:
I realized now that we're using the entry_id as unique_id for the cloud entry. That's not available until after the flow has finished. So we can't set unique_id earlier for the cloud entry with this unique id.

If we only allow a single cloud entry we could use the integration domain name as unique_id instead.

Reply:
I've used the _async_oauth_creaty_entry_ helper method and set the unique_id for cloud instances to a fixed string (unrelated to the serial numbers).